### PR TITLE
Changes for openshift as a install target

### PR DESCRIPTION
**Pull request by richardcase**
Added the `--oc` flag to indicate that the target of the install is
an OpenShift cluster. The effect of this flag is to not apply the security context part of the memcached deployment. It also then allows you to use the following flags:

* `--oc-add-memcached-scc` - this will create a memcached specific service account,  SCC and also updates the deployment to use the service account
* `--oc-add-flux-scc` - this will create a flux specific SCC.

If you don't specify the `--oc-add-*` flags then the default SCC will be applied to the flux and memcached pods. 

Still a work in progress, adding additional changes and tests.......

Relates to: #2651

### CODE REVIEW

**Comment by 2opremio** [comment]
@richardcase is this ready to review?

**Comment by richardcase** [comment]
@2opremio - yes i think it is probably ready for review

File: pkg/install/templates/memcache-account-oc.tmpl [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+users:
+    - memcached
+volumes:
+- configMap
+- downwardAPI
+- emptyDir
+- persistentVolumeClaim
+- projected
+- secret
+{{ end }}
```
Comment (2opremio) [inline pkg/install/templates/memcache-account-oc.tmpl]:
```suggestion
{{ end }}

```

File: pkg/install/templates/memcache-dep.yaml.tmpl [hunk 2]
```diff
@@ -27,8 +28,8 @@ spec:
         # - -vv    # Uncomment to get logs of each request and response.
         ports:
         - name: clients
-          containerPort: 11211
+          containerPort: 11211 {{ if not .OpenShiftTarget}}
```
Comment (2opremio) [inline pkg/install/templates/memcache-dep.yaml.tmpl]:
```suggestion
          containerPort: 11211{{ if not .OpenShiftTarget}}
```

File: pkg/install/templates/memcache-dep.yaml.tmpl [hunk 3]
```diff
@@ -27,8 +28,8 @@ spec:
...
         securityContext:
           runAsUser: 11211
           runAsGroup: 11211
-          allowPrivilegeEscalation: false
+          allowPrivilegeEscalation: false {{ end }}
```
Comment (2opremio) [inline pkg/install/templates/memcache-dep.yaml.tmpl]:
```suggestion
          allowPrivilegeEscalation: false{{ end }}
```

**Comment by 2opremio** [comment]
Apologies for taking so long to review this (Christmas vacation got on the way). What is the purpose of `oc-add-memcached-scc`and `oc-add-flux-scc`. Shouldn't they always be added when using `--oc`?

I am not familiar with SCCs so sorry if that's obvious. At the very least, I think that the Memcache service account should always be added, to make it consistent with this code

```
    spec:{{ if .OpenShiftTarget}}
      serviceAccountName: memcached{{ end }}
```

**Comment by richardcase** [comment]
> I am not familiar with SCCs so sorry if that's obvious. At the very least, I think that the Memcache service account should always be added, to make it consistent with this code

@2opremio  - they are a OpenShift specific way of giving pods specific permissions. A bit like PodSecurityPolices.  If you don't specify an explicit SCC then OpenShift will assign one of the builtin in SCCs to the pod. I kept them as optional flags for this reason but we would live without the explicit SCCs which would be easier.

**Comment by richardcase** [comment]
And maybe we can change the `--oc` flag to `--no-security-contect`?

**Comment by 2opremio** [comment]
> And maybe we can change the `--oc` flag to `--no-security-contect`?

I like that, since then we wouldn't have target-specific flags.

**Comment by richardcase** [comment]
> I like that, since then we wouldn't have target-specific flags.

Great, i'll make that change now

**Comment by richardcase** [comment]
@2opremio - i made the change for `--no-security-context`

File: pkg/install/templates/memcache-account.yaml.tmpl [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+apiVersion: v1
+kind: ServiceAccount
```
Comment (2opremio) [inline pkg/install/templates/memcache-account.yaml.tmpl]: Why do we need a specific service account for memcached? Why not use the flux service account?

File: pkg/install/install_test.go [hunk 5]
```diff
@@ -39,12 +61,22 @@ func TestFillInTemplatesAllParameters(t *testing.T) {
...
+		NoSecurityContext:  false,
+	}
+
+	manifests := testFillInTemplates(t, 6, params)
+
+	memDeploy := manifests["memcache-dep.yaml"]
+	deployment := testExtractDeployment(t, memDeploy)
 
+	container := deployment.Spec.Template.Spec.Containers[0]
+	if container.SecurityContext == nil {
```
Comment (2opremio) [inline pkg/install/install_test.go]: Please create a separate test for this `TestFillInTemplatesAllParameters` is supposed to only do that

File: cmd/fluxctl/install_cmd.go [hunk 6]
```diff
@@ -51,6 +51,7 @@ fluxctl install --git-url 'git@github.com:<your username>/flux-get-started' | ku
 	cmd.Flags().BoolVar(&opts.RegistryScanning, "registry-scanning", true,
 		"scan container image registries to fill in the registry cache")
 	cmd.Flags().StringVarP(&opts.outputDir, "output-dir", "o", "", "a directory in which to write individual manifests, rather than printing to stdout")
+	cmd.Flags().BoolVar(&opts.NoSecurityContext, "no-security-context", false, "Ensure that no security context information is added to the pod specs. Defaults to 'false'")
```
Comment (2opremio) [inline cmd/fluxctl/install_cmd.go]: I think negatives are confusing. How about `--add-security-context` and defaulting it to `true`?

**Comment by 2opremio** [comment]
BTW, we should probably do the same changes on the Helm chart

**Comment by 2opremio** [comment]
Also, we should update the documentation with this. I don't know if we already include how to install Flux for OpeShift but we should probably add/update that.

File: pkg/install/templates/memcache-account.yaml.tmpl (hunk 4 above)
Comment (richardcase) [reply pkg/install/templates/memcache-account.yaml.tmpl]: We probably don't. I will remove

File: pkg/install/install_test.go (hunk 5 above)
Comment (richardcase) [reply pkg/install/install_test.go]: Good point, i have created this test

File: cmd/fluxctl/install_cmd.go (hunk 6 above)
Comment (richardcase) [reply cmd/fluxctl/install_cmd.go]: Yes it is confusing, very good point. I have changed this.

**Comment by richardcase** [comment]
> BTW, we should probably do the same changes on the Helm chart

Ah yes, i hadn't thought of that. I have updated it now.

> Also, we should update the documentation with this. I don't know if we already include how to install Flux for OpeShift but we should probably add/update that.

The docs do need to be updated. There is #1125 about updating the docs for openshift. Would you like me to update the docs as pary of this change? Or do a separate doc PR at a later date?

**Comment by 2opremio** [comment]
No preference

File: deploy/memcache-account.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (2opremio) [inline deploy/memcache-account.yaml]: I think you forgot to remove this

Comment (richardcase) [reply deploy/memcache-account.yaml]: Yep missed that after removing the memcached service account.. Removed the file now.

**Comment by richardcase** [comment]
> No preference

I'll do a separate PR and link to #1125.

File: pkg/install/templates/memcache-dep.yaml.tmpl [hunk 8]
```diff
@@ -18,6 +18,7 @@ spec:
     spec:
       nodeSelector:
         beta.kubernetes.io/os: linux
+      serviceAccountName: flux
```
Comment (2opremio) [inline pkg/install/templates/memcache-dep.yaml.tmpl]: Do you need an explicit account to make it work on openshift? Flux needs permissions Memecached doesn't, so, the default account is a better choice.

Sorry for not writing this before.

Comment (richardcase) [reply pkg/install/templates/memcache-dep.yaml.tmpl]: Previously I had always tested it with an explicit service account. But testing it just now it doesn't need it. I will remove it and push a change now.

File: pkg/install/install_test.go [hunk 9]
```diff
@@ -1,16 +1,22 @@
...
 
 	"github.com/instrumenta/kubeval/kubeval"
 	"github.com/stretchr/testify/assert"
+
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
 )
 
-func testFillInTemplates(t *testing.T, expectedManifestCount int, params TemplateParameters) {
+func testFillInTemplates(t *testing.T, expectedNoManifestst int, params TemplateParameters) map[string][]byte {
```
Comment (2opremio) [inline pkg/install/install_test.go]:
```suggestion
func testFillInTemplates(t *testing.T, expectedNoManifests int, params TemplateParameters) map[string][]byte {
```

File: pkg/install/install_test.go [hunk 10]
```diff
@@ -24,6 +30,22 @@ func testFillInTemplates(t *testing.T, expectedManifestCount int, params Templat
 			}
 		}
 	}
+
+	return manifests
+}
+
+func testExtractDeployment(t *testing.T, data []byte) *v1.Deployment {
+	manifest := string(data)
+	manifest = strings.Replace(manifest, "---\n", "", -1)
```
Comment (2opremio) [inline pkg/install/install_test.go]: Why is this needed?

Comment (2opremio) [reply pkg/install/install_test.go]: (I think it should work without it)

File: pkg/install/install_test.go [hunk 11]
```diff
@@ -24,6 +30,22 @@ func testFillInTemplates(t *testing.T, expectedManifestCount int, params Templat
 			}
 		}
 	}
+
+	return manifests
+}
+
+func testExtractDeployment(t *testing.T, data []byte) *v1.Deployment {
```
Comment (2opremio) [inline pkg/install/install_test.go]: I would simply inline this function

File: pkg/install/install_test.go [hunk 12]
```diff
@@ -24,6 +30,22 @@ func testFillInTemplates(t *testing.T, expectedManifestCount int, params Templat
...
+	manifest := string(data)
+	manifest = strings.Replace(manifest, "---\n", "", -1)
+
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer([]byte(manifest)), 4096)
+	deployment := new(v1.Deployment)
```
Comment (2opremio) [inline pkg/install/install_test.go]:
```suggestion
	var deployment v1.Deployment
```

**Comment by 2opremio** [comment]
Minor nits, otherwise LGTM

File: pkg/install/install_test.go (hunk 9 above)
Comment (richardcase) [reply pkg/install/install_test.go]: Changed.

File: pkg/install/install_test.go (hunk 10 above)
Comment (richardcase) [reply pkg/install/install_test.go]: Its not needed. I was having another issue at the time. Removed.

File: pkg/install/install_test.go (hunk 11 above)
Comment (richardcase) [reply pkg/install/install_test.go]: I haven't inlined this as its used in 2 different test functions.

File: pkg/install/install_test.go (hunk 12 above)
Comment (richardcase) [reply pkg/install/install_test.go]: Made a change to this

**Comment by richardcase** [comment]
@2opremio - thanks for the review. I've made some minor changes based on the nits.

File: pkg/install/install_test.go (hunk 11 above)
Comment (2opremio) [reply pkg/install/install_test.go]: Ah, true. Can you rename it to `unmarshalDeployment`?

**Comment by 2opremio** [comment]
One additional nit. Could you rebase on master?

**Comment by 2opremio** [comment]
BTW, thanks a lot for coping with all my comments. This PR took way longer to merge than it should had.

**Comment by richardcase** [comment]
> One additional nit. Could you rebase on master?

All done.

**Comment by 2opremio** [comment]
Note to self: this contains both helm chart changes and `fluxctl onstall` changes so it should go both into the Flux and Chart CHANGELOGs
